libkbfs/journal_server_test.go

```diff
@@ -117,3 +122,245 @@ func TestJournalServerRestart(t *testing.T) {
+	require.NoError(t, err)
+	require.True(t, rekeyDone)
+
+	_, err = mdOps.Put(ctx, rmd)
+	require.NoError(t, err)
+
+	// Simulate a log out.
+
+	serviceLoggedOut(ctx, config)
+
+	// Get the block, which should fail.
+
+	_, _, err = blockServer.Get(ctx, tlfID, bID, bCtx)
+	require.IsType(t, BServerErrorBlockNonExistent{}, err)
+
+	// Get the head, which should be empty.
+
+	head, err := mdOps.GetForTLF(ctx, tlfID)
+	require.NoError(t, err)
+	require.Equal(t, ImmutableRootMetadata{}, head)
+
+	serviceLoggedIn(
+		ctx, jServer.log, "test_user1", config.KeybaseService(), config,
+		TLFJournalBackgroundWorkPaused)
+
+	// Get the block.
+
+	buf, key, err := blockServer.Get(ctx, tlfID, bID, bCtx)
+	require.NoError(t, err)
+	require.Equal(t, data, buf)
+	require.Equal(t, serverHalf, key)
+
+	// Get the MD.
+
+	head, err = mdOps.GetForTLF(ctx, tlfID)
+	require.NoError(t, err)
+	require.Equal(t, rmd.Revision(), head.Revision())
+}
+
+func TestJournalServerMultiUser(t *testing.T) {
+	tempdir, config, jServer := setupJournalServerTest(t)
+	defer teardownJournalServerTest(t, tempdir, config)
+
+	// Use a shutdown-only BlockServer so that it errors if the
+	// journal tries to access it.
+	jServer.delegateBlockServer = shutdownOnlyBlockServer{}
+
+	ctx := context.Background()
+
+	tlfID := FakeTlfID(2, false)
+	err := jServer.Enable(ctx, tlfID, TLFJournalBackgroundWorkPaused)
+	require.NoError(t, err)
+
+	blockServer := config.BlockServer()
+	mdOps := config.MDOps()
+	crypto := config.Crypto()
+
+	h, err := ParseTlfHandle(
+		ctx, config.KBPKI(), "test_user1,test_user2", false)
+	require.NoError(t, err)
+	uid1 := h.ResolvedWriters()[0]
+	uid2 := h.ResolvedWriters()[1]
+
+	bh, err := h.ToBareHandle()
+	require.NoError(t, err)
+
+	// Put a block under user 1.
+
+	bCtx1 := BlockContext{uid1, "", zeroBlockRefNonce}
+	data1 := []byte{1, 2, 3, 4}
+	bID1, err := crypto.MakePermanentBlockID(data1)
+	require.NoError(t, err)
+	serverHalf1, err := crypto.MakeRandomBlockCryptKeyServerHalf()
+	require.NoError(t, err)
+	err = blockServer.Put(ctx, tlfID, bID1, bCtx1, data1, serverHalf1)
+	require.NoError(t, err)
+
+	// Put an MD under user 1.
+
+	rmd1 := NewRootMetadata()
+	err = rmd1.Update(tlfID, bh)
+	require.NoError(t, err)
+	rmd1.tlfHandle = h
+	rmd1.SetRevision(MetadataRevision(1))
+	rmd1.SetLastModifyingWriter(uid1)
+	rekeyDone, _, err := config.KeyManager().Rekey(ctx, rmd1, false)
+	require.NoError(t, err)
+	require.True(t, rekeyDone)
+
+	_, err = mdOps.Put(ctx, rmd1)
+	require.NoError(t, err)
+
+	// Log in user 2.
+
+	serviceLoggedOut(ctx, config)
+
+	service := config.KeybaseService().(*KeybaseDaemonLocal)
+	service.currentUID = uid2
+	SwitchDeviceForLocalUserOrBust(t, config, 0)
+
+	serviceLoggedIn(
+		ctx, jServer.log, "test_user2", config.KeybaseService(),
+		config, TLFJournalBackgroundWorkPaused)
+
+	err = jServer.Enable(ctx, tlfID, TLFJournalBackgroundWorkPaused)
+	require.NoError(t, err)
+
+	// Put a block under user 2.
```

Comment:
First check that none of user 1's blocks are visible, at the time when user 2's journal is empty (potentially a different case than your check below).